### ISSUE TITLE
Backslash escape not handled properly by ucar.httpservices

### ISSUE THREAD

**Issue by DennisHeimbigner** [Unidata/thredds#272]
John Caron Writes:
NetcdfFile.makeFullName() backslash escapes a ".", but those aren't accepted
by HTTPFactory.Get(url)

Not sure if HTTPFactory has a bug or something else...

@Test
public void testBackslashEscaped() throws Exception {
  String url = "http://localhost:8081/thredds/cdmremote/scanCdmUnitTests/formats/hdf5/grid_1_3d_xyz_aug.h5?req=data&var=HDFEOS_INFORMATION/StructMetadata\\.0";
  try (HTTPMethod m = HTTPFactory.Get(url)) {
    int statusCode = m.execute();
    System.out.printf("status = %d%n", statusCode);
  }
}

ucar.httpservices.HTTPException: HTTPAuthUtil: illegal url:
http://localhost:8081/thredds/cdmremote/scanCdmUnitTests/formats/hdf5/grid_1_3d_xyz_aug.h5?req=data&var=HDFEOS_INFORMATION/StructMetadata\.0
	at ucar.httpservices.HTTPAuthUtil.decompose(HTTPAuthUtil.java:79)
	at ucar.httpservices.HTTPAuthUtil.urlToScope(HTTPAuthUtil.java:98)
	at ucar.httpservices.HTTPSession.<init>(HTTPSession.java:705)
	at ucar.httpservices.HTTPFactory.newSession(HTTPFactory.java:56)
	at ucar.httpservices.HTTPMethod.<init>(HTTPMethod.java:201)
	at ucar.httpservices.HTTPMethod.<init>(HTTPMethod.java:178)
	at ucar.httpservices.HTTPFactory.Get(HTTPFactory.java:120)
	at ucar.nc2.util.net.TestMisc.testBackslashEscaped(TestMisc.java:126)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)

**Comment by DennisHeimbigner** [comment]
The problem was that java.net.URI and java.net.URL
could not parse url strings that were based on
backslash escape rather than %xx escape.
The General Rule is this:
1. any url given to e.g. HTTPSession or HTTPMethod
or HTTPFactory is assumed to be the url
that is wanted on the server side.
2. #1 means that, again as a general rule, urlstrings
given to those classes should not be %xx escaped
unless that is what you want on the server side.
3. The key to understand is that
a. those classes will internally %xx encode the url string they are
given even if the url string already is %xx encoded.
b. on the server side, the %xx encoding will be performed
on the incoming url string. This means that the %xx encoded
string you originally sent is the one given to the servlet.
Bottom line: do not do %xx encoding yourself unless your server side
code is expected to see a %xx encoded url.

Anyway, the fix is to create a procedure -- HTTPUtil.parseToURI() --
that can properly create a java.net.URI object from a backslash
escaped url string. Any time you need to create a URI object,
you should do so using that procedure.

Misc. Notes:
-- Test program TestURIParse is also included.
-- In several places, I changed the use of URL to URI
so this new code could be used. The primary effect was
to replace MalFormedURLException with URISyntaxException
in many places.
-- minor unrelated fix to TestFormBuilder is also included in this pull request:
changed port numbers for the tests.

**Comment by JohnLCaron** [comment]
What version is HTTPUtil.parseToURI()  in? is that ours or in

**Comment by JohnLCaron** [comment]
doesnt appear to be in 5.0 yet?
shall i merge with master?

**Comment by JohnLCaron** [comment]
Dennis, have you started to use HTTPUtil.parseToURI(), eg in dods client?

**Comment by DennisHeimbigner** [comment]
It is in master, so merge is appropriate.
=Dennis


John Caron wrote:

**Comment by DennisHeimbigner** [comment]
Using escapes in DAP2 (dods) would, I think,
technically be a violation of the DAP2 spec
since it assumes %xx escaping rather than backslash.
If you think it is a good idea, then I can attempt
to get dap2 to support both %xx and backslash.
=Dennis


John Caron wrote:

**Comment by JohnLCaron** [comment]
So you pass a "." in a var name without escaping it?

**Comment by JohnLCaron** [comment]
so this isnt working in cdmremote, this url:

'http://localhost:8081/thredds/cdmremote/scanCdmUnitTests/formats/hdf5/grid_1_3d_xyz_aug.h5?req=data&var=HDFEOS_INFORMATION/StructMetadata%5c%2e0'

gets rejected on the server before it hits my code. not sure if its in tomcat or in spring.

im not sure what the HTTPFactory() does with it; if you say that 

"those classes will internally %xx encode the url string they are
given even if the url string already is %xx encoded."

then i assume it encodes %5c%2e0 again, then when it decodes, we are back to 

%5c%2e0

instead of "\." 

??

Im not sure, but maybe cdmremote should forego the backslash encoding like opendap ?? what is the backslash encoding for ??

**Comment by DennisHeimbigner** [comment]
I assume you mean for dap2; technically, it must
converted to %2e although I have never checked.
=Dennis

John Caron wrote:

**Comment by DennisHeimbigner** [comment]
I am getting confused.
For DAP4, backslash escape is used.
For DAP2, %xx escape is used.
For cmdremote, the choice is up to you.

Also, can you see the raw url that the servlet is getting?
=Dennis


John Caron wrote:

**Comment by JohnLCaron** [comment]
yes, im confused also. im not sure if we have any successful use of escaping variable names correctly.

i cant see what the servlet is getting at the moment, but ill try to figure out a way to do so....

**Comment by DennisHeimbigner** [comment]
The url should be accessible as HttpServletRequest.getRequestURL().
personally, I think cdmremote should use backslash escaping because
as we have seen, using %xx escaping in dap2 has caused a lot of confusion.

=Dennis


John Caron wrote:

**Comment by JohnLCaron** [comment]
1) 'http://localhost:8081/thredds/cdmremote/scanCdmUnitTests/formats/hdf5/grid_1_3d_xyz_aug.h5?req=data&var=HDFEOS_INFORMATION/StructMetadata%5c%2e0' gets passed through HTTPMethod.execute() without change, and we see that URL at the servlet filter

2) tomcat or spring decodes it, so when it gets to Controller, its now looks like

var=HDFEOS_INFORMATION/StructMetadata\.0

yay!

**Comment by JohnLCaron** [comment]
seems to work now.

**Comment by JohnLCaron** [comment]
sorry that should have been:

2) tomcat or spring decodes it, so when it gets to Controller, its now looks like

var=HDFEOS_INFORMATION/StructMetadata\.0

which gets correctly processed by ParsedSectionSpec.parseVariableSection()

**Comment by DennisHeimbigner** [comment]
ok, good that it works now.
=Dennis


John Caron wrote:
